File: 2025_02/leetcode2025-02-18.py
from typing import List
from bisect import *
from collections import defaultdict

class RangeFreqQuery:

    def __init__(self, arr: List[int]):
        self.cnt = defaultdict(list)
        for i in range(len(arr)):
            self.cnt[arr[i]].append(i)

    def query(self, left: int, right: int, value: int) -> int:
        return bisect_right(self.cnt[value], right)-bisect_left(self.cnt[value], left)

if __name__ == "__main__":
    arr = [12,33,4,56,22,2,34,33,22,12,34,56]
    rfq = RangeFreqQuery(arr)
    ques = [[1,2,4],[0,11,33]]
    for q in ques:
        print(rfq.query(q[0],q[1],q[2]))

File: leetcode2025/2025_02/leetcode2025-02-18.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class RangeFreqQuery {
private:
    unordered_map<int, vector<int>> mp;
public:
    RangeFreqQuery(vector<int>& arr) {
        for (int i=0; i<arr.size(); i++){
            mp[arr[i]].push_back(i);
        }
    }
    
    int query(int left, int right, int value) {
        int l = lower_bound(mp[value].begin(), mp[value].end(), left)-mp[value].begin();
        int r = upper_bound(mp[value].begin(), mp[value].end(), right)-mp[value].begin()-1;
        return r-l+1;
    }
};

int main(){
    vector<int> arr = {12,33,4,56,22,2,34,33,22,12,34,56};
    RangeFreqQuery rfq(arr);
    cout << rfq.query(1,2,4) << endl;
}
